physics/cooling/include/cooling/cooling.hpp

```diff
@@ -26,219 +26,109 @@
-    std::optional<int>   LWbackground_sawtooth_suppression      = std::nullopt;
-    std::optional<int>   use_volumetric_heating_rate            = std::nullopt;
-    std::optional<int>   use_specific_heating_rate              = std::nullopt;
-    std::optional<int>   use_radiative_transfer                 = std::nullopt;
-    std::optional<int>   radiative_transfer_coupled_rate_solver = std::nullopt;
-    std::optional<int>   radiative_transfer_intermediate_step   = std::nullopt;
-    std::optional<int>   radiative_transfer_hydrogen_only       = std::nullopt;
-    std::optional<int>   H2_self_shielding                      = std::nullopt;
-    std::optional<int>   dust_chemistry                         = std::nullopt;
+    Cooler();
+    ~Cooler();
+
+    void init(const double ms_sim, const double kp_sim, const int comoving_coordinates,
+              const std::optional<chemistry_data_> grackleOptions     = std::nullopt,
+              const std::optional<std::string>    grackleOptionsFile = std::nullopt,
+              const std::optional<double>         t_sim              = std::nullopt);
+
+    GlobalValues& get_global_values();
+    chemistry_data_ getDefaultChemistryData();
+
+private:
+    struct Impl;
+    std::unique_ptr<Impl> impl_ptr;
 };
 
-void initGrackle(const std::string grackle_data_file_path, const grackle_options options,
-                 const double density_units = 1.67e-24, const double length_units = 1.0, const double time_units = 1e12,
-                 const double a_units = 1., const double a_value = 1., const int comoving_coordinates = 0)
-{
-    grackle_verbose = 0;
-    // Units in cgs
-    code_units_simulation.density_units        = density_units; // m_sun / (pc * pc * pc);
-    code_units_simulation.length_units         = length_units;  // pc;
-    code_units_simulation.time_units           = time_units;    // code_time;
-    code_units_simulation.velocity_units       = code_units_simulation.length_units / code_units_simulation.time_units;
-    code_units_simulation.a_units              = a_units;
-    code_units_simulation.a_value              = a_value;
-    code_units_simulation.comoving_coordinates = comoving_coordinates;
-
-    chemistry_data_simulation = new chemistry_data;
-    set_default_chemistry_parameters(chemistry_data_simulation);
-
-    char* grackle_data_file_path_copy = new char[grackle_data_file_path.size()];
-    strncpy(grackle_data_file_path_copy, grackle_data_file_path.c_str(), grackle_data_file_path.size());
-    chemistry_data_simulation->grackle_data_file = grackle_data_file_path_copy;
-
-    chemistry_data_simulation->use_grackle = 1;
-    auto set_option                        = [](auto& grackle_data_field, const auto& optional)
+//extern template struct Cooler<double>;
+//extern template struct Cooler<float>;
+
+//! @brief Initialize Grackle chemistry arrays with default data
+    template<typename ChemistryData>
+    void initGrackleData(ChemistryData& d, size_t n)
```

Comment:
Perhaps `initChemistryData` would be a better name for this. It's not part of the `Cooler` interface to Grackle. It's init code for the Chemistry data that is managed in the sphexa compilation unit.
This can be in a separate file called `init_chemistry.hpp`.